src/back/spv/image.rs

```diff
@@ -133,6 +176,255 @@ impl<'w> BlockContext<'w> {
+                block.body.push(Instruction::image_query(
+                    spirv::Op::ImageQuerySamples,
+                    i32_type_id,
+                    count_id,
+                    image_id,
+                ));
+            } else {
+                // This is not a multisampled image reference, but
+                // level_or_sample_id is present, so it must be a
+                // level-of-detail index. Find the number of mipmap levels in
+                // this image.
+                block.body.push(Instruction::image_query(
+                    spirv::Op::ImageQueryLevels,
+                    i32_type_id,
+                    count_id,
+                    image_id,
+                ));
+            }
+
+            // Subtract one to get the largest valid value.
+            let limit_id = self.gen_id();
+            block.body.push(Instruction::binary(
+                spirv::Op::ISub,
+                i32_type_id,
+                limit_id,
+                count_id,
+                i32_one_id,
+            ));
+
+            let restricted_id = self.gen_id();
+            block.body.push(Instruction::ext_inst(
+                self.writer.gl450_ext_inst_id,
+                spirv::GLOp::UMin,
+                i32_type_id,
+                restricted_id,
+                &[level_or_sample_id, limit_id],
+            ));
+
+            restricted_id
+        });
+
+        // Obtain the image bounds, including the array element count.
+        let coordinate_bounds_id = self.gen_id();
+        match (multisampled, restricted_level_or_sample_id) {
+            (false, Some(level_id)) => {
+                // A level of detail was provided, so fetch the image size for
+                // that level.
+                let mut inst = Instruction::image_query(
+                    spirv::Op::ImageQuerySizeLod,
+                    coordinates.type_id,
+                    coordinate_bounds_id,
+                    image_id,
+                );
+                inst.add_operand(level_id);
+                block.body.push(inst);
+            }
+            _ => {
+                // No level of detail was given.
+                block.body.push(Instruction::image_query(
+                    spirv::Op::ImageQuerySize,
+                    coordinates.type_id,
+                    coordinate_bounds_id,
+                    image_id,
+                ));
+            }
+        }
+        let ones = self.write_coordinate_one(&coordinates)?;
+        let coordinate_limit_id = self.gen_id();
+        block.body.push(Instruction::binary(
+            spirv::Op::ISub,
+            coordinates.type_id,
+            coordinate_limit_id,
+            coordinate_bounds_id,
+            ones,
+        ));
+
+        let restricted_coordinate_id = self.gen_id();
+        block.body.push(Instruction::ext_inst(
+            self.writer.gl450_ext_inst_id,
+            spirv::GLOp::UMin,
```

Comment:
might be worth noting (in a comment) that using unsigned min allows us to handle negative values